### PR TITLE
[Composable FSDP] Enable HSDP

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105207
* __->__ #105206
* #105205
* #105180
* #105176

Need to pass in strategy to _init_process_group_state to enable hsdp
for composable.

Differential Revision: [D47462394](https://our.internmc.facebook.com/intern/diff/D47462394/)

### CODE REVIEW

**Review by awgu — APPROVED** [review]

File: test/distributed/_composable/fully_shard/test_fully_shard_runtime.py [hunk 1]
```diff
@@ -46,7 +46,7 @@ class TestRuntime(FSDPTest):
 
     @property
     def world_size(self) -> int:
-        return 2
+        return torch.cuda.device_count()
```
Comment (fegin) [inline test/distributed/_composable/fully_shard/test_fully_shard_runtime.py:49]: Why this changes? For people who use the internal server to run the tests, it may result in always using 8 GPUs to do the tests. Is that what we want?

Comment (awgu) [reply test/distributed/_composable/fully_shard/test_fully_shard_runtime.py:49]: Would we be able to use `CUDA_VISIBLE_DEVICES=x,y python ...` internally to work around this if needed?

Comment (rohan-varma) [reply test/distributed/_composable/fully_shard/test_fully_shard_runtime.py:49]: We should at least use a world size of 4 because we're not testing HSDP. can we workaround internally with cuda_visible_devices as andrew mentioned?

**Review by fegin — APPROVED** [review]

**Comment by huydhn** [comment]
@rohan-varma Curiously, one of the slow test `test_testing.py::TestTestingCUDA::test_cuda_assert_should_not_stop_common_distributed_test_suite_cuda` fails starting with this change https://hud.pytorch.org/pytorch/pytorch/commit/4137d6e4997f76304f478bf56f7296e9bae46fbd.  As I'm not sure they are related, could you help take a look?

cc @DanilBaibak

**Comment by huydhn** [comment]
This can be reproduced by:

```
export PYTORCH_TEST_WITH_SLOW=1
export PYTORCH_TEST_SKIP_FAST=1

pytest -v test_testing.py -k test_cuda_assert_should_not_stop_common_distributed_test_suite_cuda
```
